tutorial/source/enumeration.ipynb

```diff
@@ -0,0 +1,516 @@
+    "elbo.loss(model, config_enumerate(guide, \"sequential\"));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Parallel enumeration is cheaper but more complex than sequential enumeration, so we'll focus the rest of this tutorial on the parallel variant. Note that both forms can be interleaved.\n",
+    "\n",
+    "### Multiple latent variables <a class=\"anchor\" id=\"Multiple-latent-variables\"></a>\n",
+    "\n",
+    "We just saw that a single discrete sample site can be enumerated via nonstandard interpretation. A model with a single discrete latent variable is a mixture model. Models with multiple discrete latent variables can be more complex, including HMMs, CRFs, DBNs, and other structured models. In models with multiple discrete latent variables, Pyro enumerates each variable in a different tensor dimension (counting from the right; see [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html)). This allows Pyro to determine the dependency graph among variables and then perform cheap exact inference using variable elimination algorithms.\n",
+    "\n",
+    "To understand enumeration dimension allocation, consider the following model, where here we collapse variables out of the model, rather than enumerate them in the guide."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "model x.shape = torch.Size([3])\n",
+      "model y.shape = torch.Size([3, 1])\n",
+      "model z.shape = torch.Size([3, 1, 1])\n"
+     ]
+    }
+   ],
+   "source": [
+    "@config_enumerate(default=\"parallel\")\n",
+    "def model():\n",
+    "    p = pyro.param(\"p\", torch.randn(3, 3).exp(), constraint=constraints.simplex)\n",
+    "    x = pyro.sample(\"x\", dist.Categorical(p[0]))\n",
+    "    y = pyro.sample(\"y\", dist.Categorical(p[x]))\n",
+    "    z = pyro.sample(\"z\", dist.Categorical(p[y]))\n",
+    "    print('model x.shape = {}'.format(x.shape))\n",
+    "    print('model y.shape = {}'.format(y.shape))\n",
+    "    print('model z.shape = {}'.format(z.shape))\n",
+    "    \n",
+    "def guide():\n",
+    "    pass\n",
+    "\n",
+    "pyro.clear_param_store()\n",
+    "elbo = TraceEnum_ELBO(max_plate_nesting=0)\n",
+    "elbo.loss(model, guide);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Plates and enumeration <a class=\"anchor\" id=\"Plates-and-enumeration\"></a>\n",
+    "\n",
+    "Pyro [plates](http://docs.pyro.ai/en/dev/primitives.html#pyro.plate) express conditional independence among random variables. Pyro's enumeration strategy can take advantage of plates to reduce exponential-cost enumeration of a cartesian product down to a linear-cost enumeration over conditionally independent random variables in lock-step. This is especially important for e.g. minibatched data.\n",
```

Comment:
Be a little more verbose about what exactly the cost is exponential in?  e.g. "reduce cost of enumeration from exponential in the `batch_shape` sizes of random variables in the plate to linear" or something like that?